### PR TITLE
add custom directives UI (WP-875)

### PR DESCRIPTION
**Pull request by vsolovei-smartling**

### CODE REVIEW

**Comment by vsolovei-smartling** [comment]
image
UI preview

File: inc/Smartling/Helpers/XmlHelper.php [hunk 1]
```diff
@@ -62,33 +56,27 @@ private static function initXml()
...
-            ),
+            'smartling.placeholder_format_custom = ' . ShortcodeHelper::getMaskRegexp(),
             'smartling.placeholder_format_custom = #sl-start#.+?#sl-end#',
         ];
 
-        foreach ($additionalComments as $extraComment) {
-            $document->appendChild($document->createComment(vsprintf(' %s ', [$extraComment])));
+        foreach (explode("\n", GlobalSettingsManager::getCustomDirectives()) as $comment) {
+            if ($comment !== '') {
+                $comments[] = $comment;
```
Comment (PavelLoparev) [inline inc/Smartling/Helpers/XmlHelper.php:74]: Do we need to left/right trim comment?

File: inc/Smartling/Services/GlobalSettingsManager.php [hunk 2]
```diff
@@ -238,4 +238,21 @@ public static function setAddSlashesBeforeSavingMeta(string $value): void
...
+        return SimpleStorageHelper::get(self::SETTING_CUSTOM_DIRECTIVES, self::SETTING_CUSTOM_DIRECTIVES_DEFAULT);
+    }
+
+    public static function setCustomDirectives(string $value): void
+    {
+        if ($value === self::SETTING_CUSTOM_DIRECTIVES_DEFAULT) {
+            SimpleStorageHelper::drop(self::SETTING_CUSTOM_DIRECTIVES);
+        } else {
+            SimpleStorageHelper::set(self::SETTING_CUSTOM_DIRECTIVES, $value);
+        }
```
Comment (PavelLoparev) [inline inc/Smartling/Services/GlobalSettingsManager.php:256]: Why do we need this logic to check for empty string? Why not just set whatever is in `$value` and check for empty string in place where you use actual value?

File: inc/Smartling/Helpers/XmlHelper.php [hunk 3]
```diff
@@ -62,33 +56,27 @@ private static function initXml()
...
             'Smartling Wordpress Connector version: ' . Bootstrap::getCurrentVersion(),
             'Wordpress installation host: ' . Bootstrap::getHttpHostName(),
-            vsprintf(
-                'smartling.placeholder_format_custom = %s',
-                [
-                    ShortcodeHelper::getMaskRegexp(),
-                ]
-            ),
+            'smartling.placeholder_format_custom = ' . ShortcodeHelper::getMaskRegexp(),
             'smartling.placeholder_format_custom = #sl-start#.+?#sl-end#',
```
Comment (PavelLoparev) [inline inc/Smartling/Helpers/XmlHelper.php:69]: How does double `placeholder_format_custom` comment work? Parser takes into consideration last directive or it merges values from double directives somehow?

**Review by PavelLoparev — COMMENTED** [review]
In general, I don't get why is this PR for if yoast pattern can be placeholderised using `placeholder_format_custom` existing directive (as I get user can enter custom regexp from the settings, right?)

**Comment by vsolovei-smartling** [comment]
> In general, I don't get why is this PR for if yoast pattern can be placeholderised using `placeholder_format_custom` existing directive (as I get user can enter custom regexp from the settings, right?)

Yes, this was the idea, we discussed with @dimitrystd, sorry for not updating the issue

File: inc/Smartling/Helpers/XmlHelper.php (hunk 1 above)
Comment (vsolovei-smartling) [reply inc/Smartling/Helpers/XmlHelper.php:74]: > Do we need to left/right trim comment?

No, whitespace should make no difference

File: inc/Smartling/Helpers/XmlHelper.php (hunk 3 above)
Comment (vsolovei-smartling) [reply inc/Smartling/Helpers/XmlHelper.php:69]: > How does double `placeholder_format_custom` comment work? Parser takes into consideration last directive or it merges values from double directives somehow?

It uses the last directive, I've removed the old format from the code

Comment (vsolovei-smartling) [reply inc/Smartling/Helpers/XmlHelper.php:69]: > I don't get it. You have
> 
> ```
> 'smartling.placeholder_format_custom = ' . ShortcodeHelper::getMaskRegexp(),
> 'smartling.placeholder_format_custom = #sl-start#.+?#sl-end#',
> ```
> 
> Only last one works. Why do we still print `'smartling.placeholder_format_custom = ' . ShortcodeHelper::getMaskRegexp(),`? And did we just break shortcode support?

It's a leftover from older code. I've added the second one in https://github.com/Smartling/wordpress-localization-plugin/pull/401/files, this includes the older directive as well. While the support for shortcodes might be somewhat lacking and dated, we didn't break shortcode support today. I've added a test that passes both on master and this branch, and the shortcode handling has not changed

**Review by PavelLoparev — APPROVED** [review]

File: inc/Smartling/WP/View/ConfigurationProfiles.php [hunk 4]
```diff
@@ -263,6 +266,12 @@ class="toggleExpert"><strong><?= __('Show Expert Settings'); ?></strong></a>
                             ?>
                         </td>
                     </tr>
+                    <tr>
+                        <th><label for="<?= GlobalSettingsManager::SETTING_CUSTOM_DIRECTIVES?>"><a href="https://help.smartling.com/hc/en-us/articles/360008000893-XML#directives"><?= __('Custom directives.')?></a><?= __('XML processor will wrap each line from the text area in') ?><pre>&lt;!-- smartling.%%line%% --&gt;</pre>. <?= __('Invalid directives WILL BREAK PROCESSING.')?></label></th>
```
Comment (dimitrystd) [inline inc/Smartling/WP/View/ConfigurationProfiles.php:270]: Added space after period. Changed text.
```suggestion
                        <th><label for="<?= GlobalSettingsManager::SETTING_CUSTOM_DIRECTIVES?>"><a href="https://help.smartling.com/hc/en-us/articles/360008000893-XML#directives"><?= __('Custom directives')?></a><?= __('. The connector will wrap each line from this text area into inline directives') ?><pre>&lt;!-- smartling.%%line%% --&gt;</pre><br><?= __('Invalid directives WILL BREAK PROCESSING.')?></label></th>
```

**Review by dimitrystd — APPROVED** [review]

**Pull request merged**